src/Collateral.sol

```diff
@@ -0,0 +1,224 @@
+     * @notice REVERT: this function will revert if the collateral transfer fails.
+     */
+    function deposit(address receiver, uint256 amount) external {
+        collateralAccounts[msg.sender][receiver].balance += amount;
+        require(collateralToken.transferFrom(msg.sender, address(this), amount));
+        emit Deposit(msg.sender, receiver, amount);
+    }
+
+    /**
+     * @dev Requests to thaw a specific amount of collateral from a receivers collateral account.
+     * @param receiver Address of the receiver the collateral account is for.
+     * @param amount Amount of collateral to thaw.
+     * @notice REVERT: this function will revert if the sender receiver collateral account does
+     *                 not have enough collateral (greater than `amount`).
+     */
+    function thaw(address receiver, uint256 amount) external {
+        CollateralAccount storage account = collateralAccounts[msg.sender][receiver];
+        uint256 totalThawingRequested = account.amountThawing + amount;
+        require(account.balance >= totalThawingRequested, "Insufficient collateral balance");
+
+        // Increase the amount being thawed
+        account.amountThawing = totalThawingRequested;
+        // Set when the thaw is complete (thawing period number of seconds after current timestamp)
+        account.thawEndTimestamp = block.timestamp + thawingPeriod;
+
+        emit Thaw(msg.sender, receiver, amount, account.amountThawing, account.thawEndTimestamp);
+    }
+
+    /**
+     * @dev Withdraws all thawed collateral from a receivers collateral account.
+     * @param receiver Address of the receiver.
+     * @notice REVERT: this function will revert if the sender receiver collateral account does
+     *                 not have any thawed collateral. This function will also revert if no thawing
+     *                 period has been completed.
+     */
+    function withdraw(address receiver) external {
+        CollateralAccount storage account = collateralAccounts[msg.sender][receiver];
+        require(account.thawEndTimestamp != 0, "No collateral thawing");
+        require(account.thawEndTimestamp <= block.timestamp, "Collateral still thawing");
+
+        // Amount is the minimum between the amount being thawed and the actual balance
+        uint256 amount = account.amountThawing > account.balance ? account.balance : account.amountThawing;
+
+        unchecked {
+            account.balance -= amount; // Reduce the balance by the withdrawn amount (no underflow risk)
+        }
+        account.amountThawing = 0;
+        account.thawEndTimestamp = 0;
+        require(collateralToken.transfer(msg.sender, amount));
+        emit Withdraw(msg.sender, receiver, amount);
+    }
+
+    /**
+     * @dev Authorizes a signer to sign RAVs for the sender.
+     * @param signer Address of the authorized signer.
+     */
+    function authorizeSigner(address signer, bytes calldata proof) external {
```

Comment:
I think the idea is to add a method to revoke an authorized signer right? If so then the proofs here allow for a replay attack, since the proof is public once posted you could re-authorize a signer even after the sender has revoked. 

We can fix that by adding a nonce to the signed message. The GRT contract implements a similar thing for permits here: https://github.com/graphprotocol/contracts/blob/dev/contracts/token/GraphToken.sol#L110
Essentially we add a mapping storing nonces per sender. Then part of the signed message is the current nonce. After validating the signed message the nonce gets incremented so the same signed message can not be re-used

Reply:
I'm not sure if this is needed. The only one who would be able to revoke a signer is the sender. The reason for revoking a signer is if the sender feels the signer has been compromised. If that is the case what would be the reason for re-adding the signer? They could have just kept the signer in the first place.

Reply:
For the sender there is no reason, but if I'm an attacker that got hold of signer keys I can reuse the proof that was originally submitted to re-add the signer as a valid one. Then I could forge my own RAVs and 🎄

Reply:
Would an attacker be able to spoof `msg.sender` to be set to the senders address? Otherwise, I don't see how the attacker getting ahold of the proof would help them. The proof check uses the `msg.sender` for the call to authorizeSigner. They would only be able to add the signer as authorized to their own address.

Reply:
This conversation can be continued on #18

Reply:
Ahh good point. Yeah agreed then, there isn't much harm that could be done.

Even then, I just realized that the `verifyAuthorizedSignerProof` uses `msg.sender` as the message, so even with a stolen proof you couldn't get past the proof verification.